### ISSUE TITLE
Allow users to generate their own tokens in user settings

### ISSUE THREAD

**Issue by julianlam** [NodeBB/nodebb-plugin-write-api#2]
... and also admins to disable this ability.

**Comment by darrenxyli** [comment]
Highly suggest to automatically create token once the user registered. and BTW, how to get the user's uid by api? Thanks

**Comment by julianlam** [comment]
@darrenxyli Thanks.

You'll want to make a call to `GET /api/user/:userslug`, which is built into NodeBB, not of this plugin. (e.g. https://community.nodebb.org/api/user/julian)

**Comment by TRGH2016** [comment]
Now that 1.0 is out, any plans to spread the love to some plugins like this that haven't seen any major updates in a while?

**Comment by julianlam** [comment]
@Dynosore Sure, are you referring to this one? :smile:

**Comment by julianlam** [comment]
@Dynosore NodeBB v1.0.0 compatible version of this plugin published as v4.0.0, available w/in the hour.

**Comment by Dynosore** [comment]
@julianlam, I mean, if you had to start somewhere, this wouldn't be a bad one ;)

I'm hoping that if I look over this code long enough I can hack together a means of getting users their own API keys that can be used by external applications as a secure means of authentication against the NodeBB install, handing off certain keys. Similar to nodebb-plugin-ns-login, but without the need to pass off a username and password through service that could potentially become compromised. Taking a look at this code seemed like a good place to start with that. I figured if the repo got some updated love, I could get some ideas from there. I just started with NodeBB, so I'm still trying to figure out all the hooks and all that fun stuff.

**Comment by julianlam** [comment]
@Dynosore Right now the write token is stored in a simple hash table, which means there's not really any facility to _expire_ them, a functionality that we'd require if we wanted to open the write api to users to create their own tokens.

Would probably need to revamp the checking layer and store expiry info in a sorted set, perhaps...

**Comment by devillized** [comment]
Hi guys,


I've managed to implement auto-generating JWT on login from the standard GET "/login" path, using a custom hook that requests a token & a modified version of  @julianlam 's write-api plugin. Now, instead of a redirect, the GET "/login" call returns a JSON response with a signed JWT token and the redirect path ( ie if a token exists in params, else normal behaviour)

the code itself does its job fine - I can now sign in as usual, Get a uid-specific JWT in the response & Use that token for API calls in write-api plugin. [tested using Postman]
 
The catch -  I'm getting a "Callback timed out" warning on login & takes a pretty long time for the response to be generated. Also not sure if the code is performant. 

**TL;DR **-  code works but hackish - needs improvement for production deployment
I'm new to JS/Node in general and could use help ! Happy to send in a pull request.

**Comment by julianlam** [comment]
v4.6.0
